rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,487 @@
+slightly different query results in the short term, as long these discrepancies
+consolidate eventually.
+
+With respect to the system core, this implies that the catalog must support
+transactions when processing partition operations. For example, a compaction
+operation may remove N partitions and replace them with a new one. These
+operations break down to a sequence of deletes, followed by an insert.
+
+### Non-requirements
+
+Infrastructure approaches for hardware redundancy are considered out of scope.
+In particular, we do not consider fault tolerance at the storage layer. There
+are many different ways to provide redundant storage, e.g., RAID storage,
+distributed file systems or NAS. Cloud storage, like S3 or Azure blob storage,
+are replicated three times, and adding an additional layer of fault tolerance on
+top does not provide a lot of additional safety but considerably increases the
+cost.
+
+## Solution Proposal
+
+The most important architectural effort concerns making the catalog a
+fault-tolerant component. To this end, we propose to factor the logic that
+operates on partition metadata into a dedicated **metastore** component. As a
+first point of contact for queries, the catalog also keeps per-partition index
+structures in memory. This catalog functionality is out of scope, as it
+represents derived, immutable state from the metastore.
+
+Adjacent fault-tolerance efforts outside the catalog scope involve replication
+of components. For example, the failure domain of the read path is a single
+query, which we can deploy in a dedicated operating system process (that extends
+to a single cloud function or container execution). Likewise, we can isolate
+failure domains on the write path by spawning each data source (or pipeline) in
+its own process. This flexibility is a by-product of the actor model
+architecture that treats internal components as microservices. The spectrum of
+fate sharing is the OS process boundary. The choice highly depends on the
+specifics of the deployment environment.
+
+We briefly touch on how the mechanism for making the write and read path
+resilient, and then focus the discussion on fault tolerance within the VAST
+server node, specifically the catalog.
+
+### Resilient Write Path
+
+Data sources provide a stream of events. VAST parses them and bundles them into
+units of partitions. Assuming that we can build a fault-tolerant catalog, the
+problem of making the write path fault-tolerant reduces to ensuring exactly-once
+processing semantics on the ingest stream.
+
+If the circumstances require high-fidelity data acquisition, we can resort to an
+external data plane with exactly-once processing stream semantics, such as
+Apache Kafka. VAST must not acknowledge events from the data plane until having
+received a confirmation that the corresponding partition is durable. In the
+presence of a crash, VAST must restart from the last acknowledged position in
+the stream.
```

Comment:
Why do you think it implies that?

Reply:
How do you roll back otherwise if you store this on a partition level? Say you have two sources that ingest events of the same schema. Now one source crashes. Can you still ack the other source?

Reply:
As a result of @rdettai I revised the section on the write path implementation. Incoming data from sources must indeed now be in their own partition.

The reason for this isn't crashing sources alone (that could be solved by storing dedicated offsets  for every source that contributed to a partition), but sources that loose their state while a database node experiences a fault. The originally proposed approach would break in this case so I changed it so the catalog would discard duplicate partitions, and that only works if a partition comes from a single source.